pulsar-broker/src/main/java/org/apache/pulsar/compaction/CompactedTopicImpl.java

```diff
@@ -100,7 +100,11 @@ public void asyncReadEntriesOrWait(ManagedCursor cursor,
                                        boolean isFirstRead,
                                        ReadEntriesCallback callback, Consumer consumer) {
             PositionImpl cursorPosition;
-            if (isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId())){
+            boolean readFromEarliest = isFirstRead && MessageId.earliest.equals(consumer.getStartMessageId())
+                && (!cursor.isDurable() || cursor.getName().equals(Compactor.COMPACTION_SUBSCRIPTION)
+                || cursor.getMarkDeletedPosition() == null
+                || cursor.getMarkDeletedPosition().getEntryId() == -1L);
```

Comment:
What does this mean?

Reply:
Related discussion: https://github.com/apache/pulsar/pull/21187#discussion_r1455147185

Reply:
Just like described in the PR's description，If a durable cursor (exclude __compaction cursor) has the mark-delete position, don't read messages from the earliest.

Reply:
Sorry, I mean why we need to check `cursor.getMarkDeletedPosition().getEntryId() == -1L`?

Reply:
> Sorry, I mean why we need to check cursor.getMarkDeletedPosition().getEntryId() == -1L?

Oh, since if `markDeletedPosition.entryID == -1L` then the md-position is an invalid position, the initial md-position of the consumer will be set to it.

Reply:
Ok, I see. Could you please help add some comments to improve the code readability?

Reply:
> Ok, I see. Could you please help add some comments to improve the code readability?

@codelipenghui I add some comments and `hasValidMarkDeletePosition` method.